lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+                        _contextManager.Add(activeContext);
+                        break;
+                    case SystemState.Fail:
+                        Logger.Log(Level.Info, "Received IActiveContext event, but system is in FAIL state. Closing the context.");
+                        activeContext.Dispose();
+                        break;
+                    default:
+                        UnexpectedState(activeContext.Id, "IActiveContext");
+                        break;
                 }
             }
         }
+        #endregion IActiveContext
 
+        #region submit tasks
         /// <summary>
-        /// Specifies what to do when evaluator fails.
-        /// If we get all completed tasks then ignore the failure
-        /// Else request a new evaluator. If failure happens in middle of IMRU 
-        /// job we expect neighboring evaluators to fail while doing 
-        /// communication and will use FailedTask and FailedContext logic to 
-        /// order shutdown.
+        /// Called from ActiveContextManager when all the expected active context are received.
+        /// It changes the system state then calls SubmitTasks().
         /// </summary>
         /// <param name="value"></param>
-        public void OnNext(IFailedEvaluator value)
+        public void OnNext(int value)
         {
-            if (AreIMRUTasksCompleted())
+            Logger.Log(Level.Info, "Received event from ActiveContextManager with NumberOfActiveContexts:" + value);
+            lock (_lock)
             {
-                Logger.Log(Level.Info,
-                    string.Format("Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
-                return;
+                // When the event AllContextsAreReady happens, change the system state from WaitingForEvaluator to SubmittingTasks
+                _systemState.MoveNext(SystemStateEvent.AllContextsAreReady);
+                SubmitTasks();
             }
+        }
 
-            Logger.Log(Level.Info,
-                string.Format("Evaluator with Id: {0} failed with Exception: {1}", value.Id, value.EvaluatorException));
-            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
-            if (currFailedEvaluators > _allowedFailedEvaluators)
+        /// <summary>
+        /// This method is responsible to prepare for the task submission then call SubmitTasks in TaskManager.
+        /// It is called in both first time and recovery scenarios.
+        /// Creates a new Communication Group and adds Group Communication Operators
+        /// For each context, adds a task to the communication group.
+        /// After all the tasks are added to the group, for each task, gets GroupCommTaskConfiguration from IGroupCommDriver 
+        /// and merges it with the task configuration.
+        /// When all the tasks are added, calls TaskManager to SubmitTasks().
+        /// </summary>
+        private void SubmitTasks()
```

Comment:
Shouldn't this method move to `TaskManager` or such? We don't seem to get much benefit from the state machine in terms of code complexity here.